CRM/Core/JobManager.php

```diff
@@ -67,13 +63,39 @@ public function execute($auth = TRUE) {
 
     // it's not asynchronous at this stage
     CRM_Utils_Hook::cron($this);
-    foreach ($this->jobs as $job) {
-      if ($job->is_active) {
-        if ($job->needsRunning()) {
-          $this->executeJob($job);
-        }
+
+    // Get a list of the jobs that have completed previously
+    $successfulJobs = Job::get(FALSE)
+      ->addWhere('is_active', '=', TRUE)
+      ->addWhere('last_run_end', 'IS NOT NULL')
+      ->addOrderBy('name', 'ASC')
+      ->execute()
+      ->indexBy('id')
+      ->getArrayCopy();
+
+    // Get a list of jobs that have not completed previously.
+    // This could be because they are a new job that has not yet run or a job that is fatally crashing (eg. OOM).
+    // If last_run_end is NULL the job has never completed successfully.
+    // If last_run_end is < last_run job has completed successfully in the past but is now failing to complete.
+    $maybeUnsuccessfulJobs = Job::get(FALSE)
+      ->addWhere('is_active', '=', TRUE)
+      ->addClause('OR', ['last_run_end', 'IS NULL'], ['last_run', '<', 'last_run_end', TRUE])
```

Comment:
```suggestion
      ->addClause('OR', ['last_run_end', 'IS NULL'], ['last_run', '>', 'last_run_end', TRUE])
```

The code comment above has it right.